src/WebJobs.Script.WebHost/Security/SecretManager.cs

```diff
@@ -138,14 +138,78 @@ public virtual HostSecretsInfo GetHostSecrets()
                         }
                     };
 
-                    string secretsContent = ScriptSecretSerializer.SerializeFunctionSecrets(secrets);
-                    File.WriteAllText(secretFilePath, secretsContent);
+                    PersistFunctionSecrets(secrets, secretsFilePath);
+                }
+
+                // Read all secrets, which will run the keys through the appropriate readers
+                var keys = secrets.Keys.Select(k => ReadSecretValue(k)).ToList();
+
+                if (keys.Any(k => k.PersistenceStateStale))
+                {
+                    UpdateFunctionSecretsPersistenceState(keys, secretsFilePath);
```

Comment:
Can't you just pass "secrets" into this method, so it is symmetric with UpdateHostSecretsPersistenceState? That keeps all the unpacking isolated in that method.